xdem/coreg.py

```diff
@@ -380,12 +399,100 @@ def icp_coregistration(reference_dem: np.ndarray, dem_to_be_aligned: np.ndarray,
             }
         ]"""
         run_pdal_pipeline(pdal_pipeline, parameters=pdal_parameters)
-        aligned_dem = rio.open(output_temp_filepath).read(1)
-        aligned_dem[aligned_dem <= lowest_value] = np.nan
+        aligned_dem = rio.open(output_dem_filepath).read(1)
+
+    if metadata is not None:
+        metadata["icp_pdal"] = pdal_metadata["stages"]["filters.icp"]
+        metadata["icp_pdal"]["nmad"] = nmad
+
+    return aligned_dem, nmad
+
+
+def icp_coregistration_opencv(reference_dem: np.ndarray, dem_to_be_aligned: np.ndarray, bounds: rio.coords.BoundingBox,
+                              mask: Optional[np.ndarray] = None, max_iterations=100, tolerance=0.05,
+                              rejection_scale=2.5, num_levels=6, metadata: Optional[dict[str, Any]] = None,
+                              **_) -> tuple[np.ndarray, float]:
+    """
+    Coregister one DEM to a reference DEM.
+
+    :param reference_dem: The input array of the DEM acting reference.
+    :param dem_to_be_aligned: The input array to the DEM acting aligned.
+    :param bounds: The bounding coordinates of the reference_dem.
+    :param mask: A boolean array of areas to exclude from the coregistration.
+    :param max_iterations: The maximum amount of iterations to run ICP.
+    :param tolerance: The minimum difference between iterations after which to stop.
+    :param rejection_scale: The threshold for outliers to be considered (scale * standard-deviation of residuals).
+    :param num_levels: Number of octrees to consider. A higher number is faster but may be more inaccurate.
+    :param metadata: Optional. A metadata dictionary that will be updated with the key "icp".
+
+    :returns: The aligned DEM (array) and the NMAD error.
+
+    # noqa: DAR101 **_
+    """
+    if not _has_cv2:
+        raise AssertionError("opencv (cv2) is not available and needs to be installed.")
+    dem_to_be_aligned_unmasked = dem_to_be_aligned.copy()
+
+    resolution = reference_dem.shape[1] / (bounds.right - bounds.left)
+    points: dict[str, np.ndarray] = {}
+    x_coords, y_coords = np.meshgrid(
```

Comment:
Is it worth making a small function for it taht would be used by both pdal and opencv's coreg?

Reply:
The PDAL and opencv conversion functions are not exactly similar. In that case, the code would have to be reformatted a bit. Maybe a task for a "simplification PR"?